vllm/core/scheduler.py

```diff
@@ -932,10 +1069,19 @@ def _schedule_prefills(
             assert len(waiting_seqs) == 1, (
                 "Waiting sequence group should have only one prompt "
                 "sequence.")
+            if (partial_prefill_metadata is not None
+                    and partial_prefill_metadata.cannot_schedule(seq_group)):
+                leftover_waiting_sequences.appendleft(seq_group)
+                waiting_queue.popleft()
```

Comment:
I feel it's safer to have `waiting_queue.popleft()` in the beginning (L1062) to make sure it is removed. Ideally `waiting_queue.popleft()` shouldn't appear more than one time in this function, but it now appears many times and I'm afraid that it may introduce bugs in the future if someone forget to pop it somewhere.

Reply:
Yeah, I hear that but I also see the opposite problem: The waiting queue needs to _not_ be popped if we don't stick the sequence group in another queue (`leftover_waiting_sequences`, `ignored_sequence_groups`, or `running` in this case) otherwise we lose the sequence group completely.

It might be better to wrap these append and pops together in little methods to ensure that they're used together